documentation/modules/tracing/proc-instrumenting-producers-consumers.adoc

```diff
@@ -0,0 +1,294 @@
+
+BiFunction<String, ProducerRecord, String> producerSpanNameProvider =
+    (operationName, producerRecord) -> "CUSTOM_PRODUCER_NAME";
+
+// Create an instance of the KafkaProducer:
+KafkaProducer<Integer, String> producer = new KafkaProducer<>(senderProps);
+
+// Create an instance of the TracingKafkaProducer
+TracingKafkaProducer<Integer, String> tracingProducer = new TracingKafkaProducer<>(producer,
+        tracer,
+        producerSpanNameProvider);
+
+// Spans created by the tracingProducer will now have "CUSTOM_PRODUCER_NAME" as the span name.
+
+// Create a BiFunction for the KafkaConsumer that operates on (String operationName, ConsumerRecord consumerRecord) and returns a String to be used as the name:
+
+BiFunction<String, ConsumerRecord, String> consumerSpanNameProvider =
+    (operationName, consumerRecord) -> operationName.toUpperCase();
+
+// Create an instance of the KafkaConsumer:
+KafkaConsumer<Integer, String> consumer = new KafkaConsumer<>(consumerProps);
+
+// Create an instance of the TracingKafkaConsumer, passing in the consumerSpanNameProvider BiFunction:
+
+TracingKafkaConsumer<Integer, String> tracingConsumer = new TracingKafkaConsumer<>(consumer,
+        tracer,
+        consumerSpanNameProvider);
+
+// Spans created by the tracingConsumer will have the operation name as the span name, in upper-case.
+// "receive" -> "RECEIVE"
+----
+
+== Built-in span names for OpenTracing Decorator instrumentation
+
+When defining custom span names, you can use the following `BiFunctions` in the `ClientSpanNameProvider` class.
+If no `spanNameProvider` is specified, `CONSUMER_OPERATION_NAME` and `PRODUCER_OPERATION_NAME` are used.
+
+.BiFunctions for defining custom span names
+[%autowidth,cols="2*",options="header",stripes="none",separator=¦]
+|===
+
+¦BiFunction
+¦Description
+
+m¦CONSUMER_OPERATION_NAME, PRODUCER_OPERATION_NAME
+¦Returns the `operationName` as the span name: "receive" for consumers and "send" for producers.
+
+m¦CONSUMER_PREFIXED_OPERATION_NAME(String prefix), PRODUCER_PREFIXED_OPERATION_NAME(String prefix)
+¦Returns a String concatenation of `prefix` and `operationName`.
+
+m¦CONSUMER_TOPIC, PRODUCER_TOPIC
+¦Returns the name of the topic that the message was sent to or retrieved from in the format `(record.topic())`.
+
+m¦PREFIXED_CONSUMER_TOPIC(String prefix), PREFIXED_PRODUCER_TOPIC(String prefix)
+¦Returns a String concatenation of `prefix` and the topic name in the format `(record.topic())`.
+
+m¦CONSUMER_OPERATION_NAME_TOPIC, PRODUCER_OPERATION_NAME_TOPIC
+¦Returns the operation name and the topic name: `"operationName - record.topic()"`.
+
+m¦CONSUMER_PREFIXED_OPERATION_NAME_TOPIC(String prefix), PRODUCER_PREFIXED_OPERATION_NAME_TOPIC(String prefix)
+¦Returns a String concatenation of `prefix` and `"operationName - record.topic()"`.
+
+|===
+
+== Attribute extraction for OpenTelemetry instrumentation
```

Comment:
@alesj -- added this, but it could use some _//comments_ on what we're trying to achieve.

Reply:
Afaik, we cannot (currently) override span or operation name in existing OpenTelemetry code,
but you can add additional attributes -- this is what we're doing here.